pkg/store/cache/caching_bucket.go

```diff
@@ -0,0 +1,389 @@
+
+		k := cachingKeyObjectBlock(name, o, end)
+		keys = append(keys, k)
+		offsetKeys[o] = k
+	}
+
+	// Try to get all blocks from the cache.
+	hits, _ := cb.cache.Fetch(ctx, keys)
+
+	for _, b := range hits {
+		cb.cachedChunkBytes.Add(float64(len(b)))
+	}
+
+	if len(hits) < len(keys) {
+		if hits == nil {
+			hits = map[string][]byte{}
+		}
+
+		err := cb.fetchMissingChunkBlocks(ctx, name, startOffset, endOffset, offsetKeys, hits, lastBlockLength)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return newBlocksReader(cb.config.ChunkBlockSize, offsetKeys, hits, offset, length), nil
+}
+
+type rng struct {
+	start, end int64
+}
+
+// Fetches missing blocks and stores them into "hits" map.
+func (cb *CachingBucket) fetchMissingChunkBlocks(ctx context.Context, name string, startOffset, endOffset int64, cacheKeys map[int64]string, hits map[string][]byte, lastBlockLength int) error {
+	// Ordered list of missing sub-ranges.
+	var missing []rng
+
+	for off := startOffset; off < endOffset; off += cb.config.ChunkBlockSize {
+		if hits[cacheKeys[off]] == nil {
+			missing = append(missing, rng{start: off, end: off + cb.config.ChunkBlockSize})
+		}
+	}
+
+	missing = mergeRanges(missing, 0) // Merge adjacent ranges.
+	// Keep merging until we have only max number of ranges (= requests).
+	for limit := cb.config.ChunkBlockSize; cb.config.MaxChunksGetRangeRequests > 0 && len(missing) > cb.config.MaxChunksGetRangeRequests; limit = limit * 2 {
+		missing = mergeRanges(missing, limit)
+	}
+
+	var hitsMutex sync.Mutex
+
+	// Run parallel queries for each missing range. Fetched data is stored into 'hits' map, protected by hitsMutex.
+	eg, nctx := errgroup.WithContext(ctx)
+	for _, m := range missing {
+		m := m
+		eg.Go(func() error {
+			r, err := cb.bucket.GetRange(nctx, name, m.start, m.end-m.start)
+			if err != nil {
+				return errors.Wrapf(err, "fetching range [%d, %d]", m.start, m.end)
+			}
+			defer runutil.CloseWithLogOnErr(cb.logger, r, "fetching range [%d, %d]", m.start, m.end)
+
+			for off := m.start; off < m.end && nctx.Err() == nil; off += cb.config.ChunkBlockSize {
+				key := cacheKeys[off]
+				if key == "" {
+					return errors.Errorf("fetching range [%d, %d]: caching key for offset %d not found", m.start, m.end, off)
+				}
+
+				// We need a new buffer for each block, both for storing into hits, and also for caching.
+				blockData := make([]byte, cb.config.ChunkBlockSize)
+				n, err := io.ReadFull(r, blockData)
+				if err == io.ErrUnexpectedEOF && (off+cb.config.ChunkBlockSize) == endOffset && n == lastBlockLength {
+					// Last block can be shorter.
+					err = nil
+					blockData = blockData[:n]
+				}
+				if err != nil {
+					return errors.Wrapf(err, "fetching range [%d, %d]", m.start, m.end)
+				}
+
+				hitsMutex.Lock()
+				hits[key] = blockData
+				hitsMutex.Unlock()
+
+				cb.fetchedChunkBytes.Add(float64(len(blockData)))
+				cb.cache.Store(nctx, map[string][]byte{key: blockData}, cb.config.ChunkBlockTTL)
```

Comment:
Due to how the `mergeRanges`  work, this block could already be in `hits`. In such case, we should definitely not restore it in the cache (it's a waste of resources).

I'm wondering what we should do with the `fetchedChunkBytes` metric. On one side it's data we actually re-fetched from the storage so increasing it could be correct. On the other side if you expect `fetchedChunkBytes + cachedChunkBytes` to be equal to the total size of data loaded from chunks then increasing it here is not correct. What do you think if we split `fetchedChunkBytes` into two: "fetched" and "refetched" (the latter would be the total number of bytes we wasted loading just to keep the number of storage API calls low)?

Reply:
I will add counter for "refetched" bytes.